File: dsa-problems/TreesDP/solutions/javascript/TDP-007-tree-independent-set-distance2.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];
rl.on("line", (line) => {
  lines.push(line);
}).on("close", () => {
  solve();
});

function solve() {
  let idx = 0;
  const n = parseInt(lines[idx++]);

  const weight = [0, ...lines[idx++].split(" ").map(Number)];

  const graph = Array.from({ length: n + 1 }, () => []);

  for (let i = 0; i < n - 1; i++) {
    const [u, v] = lines[idx++].split(" ").map(Number);
    graph[u].push(v);
    graph[v].push(u);
  }

  const dp = Array.from({ length: n + 1 }, () => [0, 0, 0]);

  function dfs(u, parent) {
    dp[u][0] = 0;
    dp[u][1] = 0;
    dp[u][2] = weight[u];

    let sumWithoutSelected = 0;
    let maxGain = -Infinity;

    for (const v of graph[u]) {
      if (v === parent) continue;

      dfs(v, u);

      const bestNotSelected = Math.max(dp[v][0], dp[v][1]);
      sumWithoutSelected += bestNotSelected;

      const gain = dp[v][2] - bestNotSelected;
      maxGain = Math.max(maxGain, gain);

      dp[u][2] += dp[v][0];
    }

    dp[u][0] = sumWithoutSelected;

    if (maxGain > -Infinity) {
      dp[u][1] = sumWithoutSelected + maxGain;
    }
  }

  dfs(1, -1);

  const result = Math.max(dp[1][0], dp[1][1], dp[1][2]);
  console.log(result);
}

File: dsa-problems/TreesDP/solutions/java/TDP-007-tree-independent-set-distance2.java
import java.util.*;

public class Distance2IndependentSet {
    static List<Integer>[] graph;
    static long[] weight;
    static long[][] dp;
    static int n;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();

        weight = new long[n + 1];
        for (int i = 1; i <= n; i++) {
            weight[i] = sc.nextLong();
        }

        graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < n - 1; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph[u].add(v);
            graph[v].add(u);
        }

        dp = new long[n + 1][3];

        dfs(1, -1);

        long result = Math.max(dp[1][0], Math.max(dp[1][1], dp[1][2]));
        System.out.println(result);

        sc.close();
    }

    static void dfs(int u, int parent) {
        dp[u][0] = 0;
        dp[u][1] = 0;
        dp[u][2] = weight[u];

        long sumWithoutSelected = 0;
        long maxGain = Long.MIN_VALUE;

        for (int v : graph[u]) {
            if (v == parent) continue;

            dfs(v, u);

            long bestNotSelected = Math.max(dp[v][0], dp[v][1]);
            sumWithoutSelected += bestNotSelected;

            long gain = dp[v][2] - bestNotSelected;
            maxGain = Math.max(maxGain, gain);

            dp[u][2] += dp[v][0];
        }

        dp[u][0] = sumWithoutSelected;

        if (maxGain > Long.MIN_VALUE) {
            dp[u][1] = sumWithoutSelected + maxGain;
        }
    }
}
